Split polygons at dateline

sindizzy opened the pull request:
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/NetTopologySuite/NetTopologySuite/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository.
<!--These follow strict Stylecop rules :cop:.-->
- [x] I have provided test coverage for my change (where applicable)

### Description
Presents a technique for converting coordinates to a polygon. With the coordinates all in one hemisphere it is a
fairly simple process. However, when the geometry crosses the dateline it becomes a more challenging problem.
This routine attempts to do both. That is to say, it converts coordinates to a poly geometry regardless if the
geometry traverses the dateline or not. If the geometry does cross the dateline then it is split up into multiple
parts. This benefits the user in two ways: first, any GIS will be able to display the geometry in a proper way regardless of map projection and two, the GIS will be able to perform geoprocessing functions on the geometry. Note that feature classes do
not support features with multiple parts that share the same edge. Thus, it is left up to the user on how to
commit the geometries to a feature class. Suggestions are to save the multiple geometries as individual features
or to slightly alter the edges of each geometry so as to save all parts as a single multi-part feature. For the
latter option, the algorithm provides an input to trim edges of parts so that they can be used in a single feature.

The algorithm arose from attempting to convert worldwide airspaces (which can be found in any online aviation db) to polygons. Some of the coordinates result in polygons traversing the dateline and the result is ugly. That is because the GIS is not inherently aware that the polygons cross the dateline. This is what a standard conversion will look like:

![RegMap_StdProc](https://user-images.githubusercontent.com/5298460/148291970-df105104-843b-4351-8054-b1e7f899a634.png)

The goal is to split those special polygons so that the airspaces are properly represented:

![RegMap_SpcProc](https://user-images.githubusercontent.com/5298460/148292088-53a79ec5-278d-4c8d-85d2-c1c02270ee00.png)

The PR includes a sample input with three processing options. If the item crosses the dateline then you will **always** get back multiple geometries no matter what options you use. The difference is what the dev intends to do with the geometries. The sample input will produce the results discussed below. For easy viewing the map projection has been set to Plate Carrée with the central meridian at 180 degrees (the dateline). The PR is also heavy on the debugs and I can remove those once the NTS devs can take a look at the initial proposal.

If the dev intends to add them as a **single part per feature** then this is what you will get with the standard options:

![dateline_sp](https://user-images.githubusercontent.com/5298460/148292756-9f64fd7f-07f8-44bc-bf7f-49f173ea27b0.png)

If the dev intends to add them as **multiple parts per feature** then its a bit more complicated. Through research and observation it was noted that existing datasets representing worldwide polygons already deal with the multi-part issue by trimming one of the polygons ever so slightly such that the two parts do not mathematically share an edge. From my research this trim gap appears to be connected to the feature class tolerance. The trim gap is a variable that the user must provide but should be made as small as possible. For demonstration purposes the trim gap has been made quite large and this is what it would look like with advanced options:

![dateline_mp](https://user-images.githubusercontent.com/5298460/148292864-fd7b0e61-9c26-49d0-8fbf-90d319eae3e9.png)

Abel

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NetTopologySuite/NetTopologySuite?pullRequest=580) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/NetTopologySuite/NetTopologySuite?pullRequest=580) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: DGuidi<br/>:white_check_mark: sindizzy<br/>:x: Perez<br/><hr/>**Perez** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/NetTopologySuite/NetTopologySuite?pullRequest=580) it.</sub>

DGuidi commented on src/NetTopologySuite/NetTopologySuite.csproj:
```diff
@@ -31,7 +31,6 @@
   </ItemGroup>
 
   <ItemGroup Condition=" '$(EnableApiCompat)' == 'true' ">
-    <PackageReference Include="Microsoft.DotNet.ApiCompat" Version="6.0.0-beta.21159.11" PrivateAssets="All" />
```
I see a potential problem here, because `EnableApiCompat == True` but the reference to the package was removed.
Honestly I don't even know how this package is used, but I think it should be restored

sindizzy replied:
I wrote an article on my issue here https://github.com/NetTopologySuite/NetTopologySuite/issues/577
I didn't know how to fix it. It was suggested I remove a package and recompile. That seemed to work.

DGuidi replied:
My doubt is about pushing the package removal, I suggest to re-add the reference and then investigate the problem in #577

FObermaier replied:
This has to go away. If it does not work in your environment, you have to fix your environment.
If you want to turn it off temporarily, change `EnableApiCompat` to false (see 1st PropertyGroup)

sindizzy replied:
Yes but I don't know how to go about fixing my environment. I'm not even sure where to start. I did a google search and not much came up. I'm just unfamiliar with ApiCompat and how it works.

airbreather replied:
> we have a fix for this

Not sure about that one, since the NuGet.config file does actually exist in this project... the question is what's causing it to get ignored?

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+    public class SplitPolygonAtDateline
+    {
+        [STAThread]
+        public static void main()
```
maybe this code can be moved to a unit test (along with other sample code, possibly?) I know that other "Technique" items uses this parrern (inherited from JTS code) but basically this means that this code isn't tested.

sindizzy replied:
I haven't done any unit testing in any of my projects so am a complete rookie at that. As you mentioned I just followed the pattern of testing in other routines in the Technique area.

DGuidi replied:
I will help you with that

DGuidi replied:
@sindizzy  added the [unit test code](https://github.com/sindizzy/NetTopologySuite/commit/ed8b5368436a496200b94f3b7c2e6e6df2c86e75), hope to not have done a mess.
This is just a sample, please feel free to add your own asserts and other checks.

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+            // ========================================================================================
+            // inputs
+            // ========================================================================================
+            string wktPrj = "GEOGCS['WGS 84',DATUM['WGS_1984',SPHEROID['WGS 84',6378137,298.257223563,AUTHORITY['EPSG','7030']],AUTHORITY['EPSG','6326']]," +
```
From what I understand, the projection string is used just ot check if the data is in a GEOGCS format. To me this check is useless, because we are not checking the geometry but a separate parameters. Maybe a check of the geometry values (inside the valid range of values) can be slower but useful?

sindizzy replied:
I suppose I can do that but say if someone has coordinates in UTM that go from 50 to 170 meters. That will pass the geographic check as we are only checking values and not units. My thought was that I didn't want users to just hand over their coordinates hoping it would work with any projection. The easiest way, for me at least, was to force them to provide a projection string.

DGuidi replied:
I see your point, but maybe it's something related to the documentation of the method/api. my2cents

sindizzy replied:
Absolutely. I strived to add quite a bit of a description buttttttt there will be many that don't read it and just dive in. :)

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+        /// </remarks>
+        public static Geometry ToPolyExOp(List<Coordinate> inpCoords, string inpProj4Wkt, OgcGeometryType outType, double outTrimGap, double outDensifyResolution, bool isAttFixOutInvPolygons)
+        {
+            Debug.Print("$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$");
```
these "debug.print" should be removed from the main methods.

sindizzy replied:
Yes, thank you. Per my main write-up "The PR is also heavy on the debugs and I can remove those once the NTS devs can take a look at the initial proposal."

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+            // ###################################################################################################################
+            if (inpCoords == null)
+            {
+                throw new Exception("Null input coordinates.");
```
please use "ArgomentNullException" instead of generic "Exception"

sindizzy replied:
Will do.

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+            {
+                case OgcGeometryType.LineString:
+                    if (inpCoords.Count < 2)
+                        throw new Exception("Input list has fewer than 2 coordinates. To create a polyline, the minimum is 2 coordinates.");
```
ArgumentOutOfRange (of maybe InvalidOperationException?) here

sindizzy replied:
I think I agree with your assessment here and will change my code. However, on my work related code I like to tell the user exactly what is wrong. At times I've encountered the ArgumentOutOfRange exception and wondered "ok what is the range then?".

DGuidi replied:
you can leave the message. I see your point and ArgumentException can be used too. 
I must admit that, once you don't use the base `Exception` class, the proper exception to use can be arguable , so feel free to use the exception class you want. Often I use `InvalidOperationException`, but not sure I can be taken as an example :)

sindizzy replied:
Sure thing. I am learning on the go.

sindizzy replied:
I think I can use something like:

`throw new ArgumentOutOfRangeException(nameof(count), count, "Must be non-negative.");`

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+                        throw new Exception("Input list has fewer than 3 coordinates. To create a polygon, the minimum is 3 coordinates.");
+                    break;
+                default:
+                    throw new Exception("Unsupported output geometry type.");
```
NotSupportedException if the operation can not be implemented, or NotImplementedException if you plan to extend the code to support these operations in the future

sindizzy replied:
got it. thanks.

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+        /// https://desktop.arcgis.com/en/arcmap/10.3/manage-data/editing-fundamentals/creating-and-editing-multipart-polygons.htm
+        /// https://desktop.arcgis.com/en/arcmap/10.3/guide-books/map-projections/what-happens-to-features-at-180-dateline-.htm
+        /// </remarks>
+        public static Geometry ToPolyExOp(List<Coordinate> inpCoords, string inpProj4Wkt, OgcGeometryType outType, double outTrimGap, double outDensifyResolution, bool isAttFixOutInvPolygons)
```
`IList` can be used, I think

sindizzy replied:
How different is using ListOf vs IList? I'm curious as to the pros and cons. 
Thanks.

DGuidi replied:
using `IList`, the caller can use as input any implementation if the [IList interface](https://docs.microsoft.com/en-us/dotnet/api/system.collections.ilist?view=net-6.0), like [arrays](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/)

FObermaier replied:
I'd prefer `inpCoordinates` to be of type `CoordinateSequence`. Maybe this whole function should be encapsulated in a class implementing `IEntireCoordinateSequenceFilter` invoked by `Geometry.Apply(filter)`

FObermaier replied:
Or have an overload of the function taking a `Geometry` as an argument. In that case `IGeometryFilter` would be an option.

sindizzy replied:
Ok let me take a look at the NTS codebase and see how CoordinateSequence is used.

sindizzy replied:
So how would the Geometry as an input work?
1. The assumption that any geometry can be broken down to a coordinate list and then the user selects what kind of output comes from those coordinates? -OR-
2. The assumption that if I provide a Polygon then I want output as polygon. If I provide a polyline then I want polyline for output? What if the user inputs other non-supported geometry types?

What makes the most sense?

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+                    Debug.Print("cutterPgon isValid = {0}", valOp2.IsValid);
+
+                    // perform an Erase to split the polygon with a polygon. this consists of an Intersection and then a Difference.
+                    //var intx = pPolygonShifted.Boundary.Intersection(cutterPgon);
```
Old commented code that can be removed?

sindizzy replied:
yep. will remove after the review is done. I'm an over-debugger I have to confess. I wanted to double check every output so  have tons of debugs. :)

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+                    Debug.Print("resGeoms were fixed");
+
+                    // method 2
+                    //var pm = new PrecisionModel(10000000000.0);
```
Old code that can be removed? or it can be handled using a specific methofd parameter?

sindizzy replied:
well no. the routine to fix invalid geometries resulted from some research and eventually this article: https://github.com/NetTopologySuite/NetTopologySuite/issues/124
From that discussion it was noted that Buffer does not always work. So I added the second method in case I wanted to  add as an option later. I could probably make the input an Enum rather than a Boolean. So I would have 

```
    public enum attFixOutInvPolygons: int
    {
        FixNone= 0,
        FixViaBuffer=1,
        FixViaPrecision=2,
    }
```

DGuidi replied:
I like the enum idea

sindizzy replied:
Im adding it as my next commit.

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+        /// https://desktop.arcgis.com/en/arcmap/10.3/manage-data/editing-fundamentals/creating-and-editing-multipart-polygons.htm
+        /// https://desktop.arcgis.com/en/arcmap/10.3/guide-books/map-projections/what-happens-to-features-at-180-dateline-.htm
+        /// </remarks>
+        public static Geometry ToPolyExOp(List<Coordinate> inpCoords, string inpProj4Wkt, OgcGeometryType outType, double outTrimGap, double outDensifyResolution, bool isAttFixOutInvPolygons)
```
"out" parameters (all the parameters following "inpProj4Wkt") can be moved to a single "parameter class", just for the sake of the readability

sindizzy replied:
Question, is there a number of parameters where you move to a parameter class? Obviously if you have 30 params you don't want those all in the function call. If I have 3 or 4, I typically leave it in the function call as-is. Looking at the NTS core code there are various calls with 5 even 6 params right in the function call. So I'm wondering if this change is needed? I'm open to it.

DGuidi replied:
Simple a suggestion, just because there are several "out" parameters, that _maybe_ can be grouped for the sake of the readability.
Totally unrelated, see something like JsonConvert.Deserialize(jsonString, jsonDeserializationParamsClass).

sindizzy replied:
Ok I will take a loo. Thanks.

DGuidi commented:
I just added some "formal" comments about the code, I don't know so much about the algorithm but I think you can add few unit test methods to show how to use the class to obtain different kind of results.
thanks for the addition anyway.

euangordon commented:
@sindizzy This looks fantastic and fingers crossed it works well. 

One question: We have many polygons crossing the 180 meridian. With this change will we be able to intersect with those polygons?

I will try and test this out myself, just wondering if you know already

DGuidi commented:
> This looks fantastic and fingers crossed it works well

@euangordon can you suggest some more unit tests?

sindizzy commented:
> I just added some "formal" comments about the code, I don't know so much about the algorithm but I think you can add few unit test methods to show how to use the class to obtain different kind of results. thanks for the addition anyway.

Thanks I will spend some time this week to clean up the code and fix the commit authoring.

As for the tests, I added 3 tests on how to use the routine. I followed other routines in the same area ie test\NetTopologySuite.Samples.Console\Technique.

DGuidi commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+        /// A Geometry class that contains the result of converting the coordinates into a Polyline or Polygon. The result could be multiple
+        /// geometries.
+        /// </returns>
+        private static Geometry ToPolyExOp(List<Coordinate> coordsInp, string proj4WktInp)
```
should be made public, I suppose

DGuidi commented:
not sure about how the unit tests in the XUniRunner are now failing, I see the same problem in my dev machine

FObermaier commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+        }
+
+        /// <summary>
+        /// This routine takes a list of Coordinates and attempts to convert them to a poly geometry (polyline or polygon). Special care is
```
What do I have to do when I have a `[Multi]LineString` that crosses the dateline?

sindizzy replied:
Right now I only implemented polygon. That's what I so very needed and I wanted to make sure that was solid before I started experimenting with polylines.

FObermaier commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+            // based on a geographic coordinate system. if the user wishes to use a projected coordinate system then they will have
+            // to reproject the points first and then use this process.
+            // ###################################################################################################################
+            if (inpProj4Wkt == null || !inpProj4Wkt.StartsWith("GEOGCS"))
```
This check (and the parameter) can be removed. One is only attempting to fix dateline issues when one has geographic coordinates. Maybe a test on the extent of the input coordinates as a replacement? In any case, if the width of the envelope is <180, there can't be a dateline crossing.

DGuidi replied:
same suggestion I provided, see [reply](https://github.com/NetTopologySuite/NetTopologySuite/pull/580#discussion_r785487001)

sindizzy replied:
Users will no doubt try to use it with any projection and realize it doesn't work. I was trying to be proactive about feedback that it is broken when using UTM or other non-geographic projections. But as I said before, what if the UTM projection goes from 50 to 170 meters?

FObermaier commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+        /// </summary>
+        /// <param name="coordsInp">The input list of coordinates.</param>
+        /// <returns>True if the coordinates as a polygon cross the dateline, False otherwise.</returns>
+        private static bool IsPolygonCrossesDatelineEx(List<Coordinate> coordsInp)
```
Again, this should work on a `CoordinateSequence`

sindizzy replied:
I will scope that out.

sindizzy replied:
This is a function internal to the process. I am converting the CoordnateSequence to a List<Coordinate> anyway since I require a bunch of List functionality.

DGuidi replied:
I think you are using onlu `List.Exists` method, that can be re-implemented easily for coordinate sequences

sindizzy replied:
I use more List functions than that I believe. If the CoordinateSequence does not provide IEnumerable then what are the alternatives?

FObermaier commented on test/NetTopologySuite.Samples.Console/Technique/SplitPolygonAtDateline.cs:
```diff
@@ -0,0 +1,802 @@
+            return cutterPolygon;
+        }
+
+        private static void DebugPrintCoords(List<Coordinate> coords, string hdr)
```
Give these `Debug...` functions a return type of `string` use a `StringBuilder` to assemble the output and call the functions using `System.Diagnostics.Debug.WriteLine(~Debug~...)`

sindizzy replied:
I will probably take all that out here after Im done with the other suggestions. It was more for me than anything else. Thank you for the feedback.

FObermaier reviewed (changes requested):
Instead of shifting all longitudes we could only shift the longitudes of geometries that cross the dateline by 360. In that case we have can use 2 constant intersection polygons and translate the geometries in the range >180 <=540 by -360

Just an idea.

sindizzy commented:
> Instead of shifting all longitudes we could only shift the longitudes of geometries that cross the dateline by 360. In that case we have can use 2 constant intersection polygons and translate the geometries in the range >180 <=540 by -360

Yes I think you suggested that way back with my first post https://github.com/NetTopologySuite/NetTopologySuite/issues/330. I also have a todo in my function description. I think by the time I realized that may have been a better option I was already deep in the weeds with the current implementation. It shouldn't be too hard for me to do what you describe here and it actually might be a more logical flow.

sindizzy commented:
> Or have an overload of the function taking a Geometry as an argument. In that case IGeometryFilter would be an option.

Yes I think I can do that. Thanks.

sindizzy commented:
Ok I tried to change the commit authors to my personal account but I'm not sure that it worked. At this point I wonder if I should just try with another brand new PR.

FObermaier commented:
@sindizzy, don't worry about the CLA.

FObermaier commented:
Please have a look at https://gist.github.com/FObermaier/ba25a5b6ccc98cb86efbbd485d17bc01

sindizzy commented:
> Please have a look at https://gist.github.com/FObermaier/ba25a5b6ccc98cb86efbbd485d17bc01

What am I looking for?

sindizzy commented:
> Please have a look at https://gist.github.com/FObermaier/ba25a5b6ccc98cb86efbbd485d17bc01

Is that the final routine?

FObermaier commented:
@sindizzy, that is how I would do it, does it work for you?

sindizzy commented:
> @sindizzy, that is how I would do it, does it work for you?

Let me test it. Can I give you some feedback later in the week?

sindizzy commented:
Also on line 87 I am not certain I understand the need to union the results. if we have a multi part polygon and we split it along the dateline then the output may be something we don't expect after the union. I would have to give this some more thought.

FObermaier commented:
@sindizzy the example posted was not meant to replace your solution, it was meant to demonstrate how to work with `CoordinateSequence`s and `I[Entire]CoordinateSequenceFilter`s. I have next to no knowledge about dateline crossing caveats, everything I have to work with is planar or far enough from the dateline.

sindizzy commented:
Ahhhh gotcha. Ok let me use that as a base for my logic.

sindizzy commented:
@FObermaier Ok based on your suggestion I added a DatelineUtility class to my PR. I think I understand your preferred workflow now. 

I did change the main filter to
`private class IsCrossesDatelineFilter : IGeometryFilter`

This is because if a user passes a multi-part polygon then each polygon part could exist wholly in one hemisphere. If we use the CoordinateSequence then that could give us a false positive. Before I continue with this code I wanted to see if I am on the right track here. The filter cycles through each part in the geometry and then through the Coordinate array. I don't use the CoordinateSequence so wondering if that will pose a problem?

sindizzy commented:
Alright, I added a CoordinateSequence filter that is called from the Geometry filter. This should account for multi-part geometries as the input. If we **_just_** do a CoordinateSequence then the coordinate sweep could jump from one part to another and we could get a false positive.

FObermaier commented on test/NetTopologySuite.Samples.Console/Technique/DatelineUtility.cs:
```diff
@@ -0,0 +1,233 @@
+                // since we are filtering per geometry we could have multiple parts. so we must test each part individually for traversing
+                // dateline. if we were to test the CoordinateSequence with a IEntireCoordinateSequenceFilter filter then we could get false
+                // positives (ex: two islands on each side of the dateline).
+                System.Diagnostics.Debug.Print("num geoms={0}", geometry.NumGeometries);
```
I'd argue that if a single `Polygon` of a `MultiPolygon` crosses the dateline you should consider the whole geometry to cross the dateline.

FObermaier commented on test/NetTopologySuite.Samples.Console/Technique/DatelineUtility.cs:
```diff
@@ -0,0 +1,233 @@
+                    //    else if (currX > -180d) wasGT = true;
+                    //}
+
+                    bool hasLonsLt180 = part.Coordinates.Any(c => c.X < 180);
```
Don't work on `Geometry.Coordinates`, that is potentially a bottleneck. For this check you iterate over the whole array twice + you might need to build the array of coordinates twice. Use `IEntireCoordinateSequenceFilter` to do that check.

FObermaier commented:
You don't need to split up into seperate filters. `IEntireCoordinateSequenceFilter` runs on all `CoordinateSequence`s of a geometry's components, as long as `IsDone` is false.

Note: Only `Point` and `LineString` geometry classes have `CoordinateSequence` members. So if you have a `Polygon` the `IEntireCoordinateSequenceFilter` will be applied to the shell and holes (which are `LinearRing`s (a special form of a `LineString`)) individually. Same for all the `Multi...` geometries and `GeometryCollection`s.

sindizzy commented:
@FObermaier Ahhhhh I think I got it now. Thanks for taking time to teach me how this works. I went back to **_only_** the IEntireCoordinateSequenceFilter and did a commit. I think this is what you were talking about so when you get a chance please take a look.

sindizzy commented:
I noticed you updated your routine so I started over with DatelineUtility2. Added my logic and the geographic CRS list. I will need to study it further to understand certain parts.

airbreather commented:
So, I haven't really been following this closely at all, but I do want to at least point out that the term "dateline" probably shouldn't show up in the code.  I don't mind referring to it that way informally, but the term "antimeridian" is a more accurate term to use when referring to the thing that causes problems for NTS, since the International Date Line deviates from the true meridian in several places in order to avoid chopping countries in half.

euangordon commented:
Hi All, what is the status of this? Is there any chance it will get merged in?

sindizzy commented:
I was working on it and I lost steam. I'll pick it up again soon.

euangordon commented:
> I was working on it and I lost steam. I'll pick it up again soon.

Thanks @sindizzy we did try and use your fork but didn't manage to get it working. Hopeing that once it is in the release build we will manage to get it working.

As a workaround, I am working on splitting shapes that cross the dateline, but it painful.

FObermaier commented:
> Hi All, what is the status of this? Is there any chance it will get merged in?

In the meantime you can try https://gist.github.com/FObermaier/ba25a5b6ccc98cb86efbbd485d17bc01

sindizzy commented:
> > I was working on it and I lost steam. I'll pick it up again soon.
> 
> Thanks @sindizzy we did try and use your fork but didn't manage to get it working. Hopeing that once it is in the release build we will manage to get it working.
> 
> As a workaround, I am working on splitting shapes that cross the dateline, but it painful.

Yes I know how painful it is. You can try the algorithm here https://github.com/NetTopologySuite/NetTopologySuite/discussions/496 That's where the discussion started.

sindizzy commented:
I had this as a side project and am going to see if I can get back to refactoring it. I've spent the last couple of years transitioning from VB to C#. I've also upgraded my own algorithm slightly.

FObermaier commented:
Closed with https://github.com/NetTopologySuite/NetTopologySuite/pull/794
